### PR TITLE
[BEAM-9022] publish Spark job server Docker image

### PR DESCRIPTION
**Pull request by ibzib**
**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

File: release/src/main/scripts/build_release_candidate.sh [hunk 1]
```diff
@@ -236,6 +236,9 @@ if [[ $confirmation = "y" ]]; then
     ./gradlew ":runners:flink:${ver}:job-server-container:dockerPush" -Pdocker-tag="${RELEASE}_rc${RC_NUM}"
   done
 
+  echo '-------------Generating and Pushing Spark job server image-------------'
+  ./gradlew "./gradlew :runners:spark:job-server:container:dockerPush" -Pdocker-tag="${RELEASE}_rc${RC_NUM}"
```
Comment (Hannah-Jiang) [inline release/src/main/scripts/build_release_candidate.sh]: Don't we need to specify spark version here?

File: website/src/contribute/release-guide.md [hunk 2]
```diff
@@ -717,6 +717,12 @@ for ver in "${FLINK_VER[@]}"; do
 done
 ```
 
+* Build Spark job server image and push to DockerHub.
+
+```
+./gradlew "./gradlew :runners:spark:job-server:container:dockerPush" -Pdocker-tag="${RELEASE}_rc${RC_NUM}"
```
Comment (Hannah-Jiang) [inline website/src/contribute/release-guide.md]: `./gradlew` is repeated.

Comment (ibzib) [reply website/src/contribute/release-guide.md]: Whoops, good catch.

File: release/src/main/scripts/build_release_candidate.sh (hunk 1 above)
Comment (ibzib) [reply release/src/main/scripts/build_release_candidate.sh]: No, the Spark runner only supports one Spark version per Beam release (currently [2.4.5](https://github.com/apache/beam/blob/800754399baa939232b9863b5104111c638e88aa/buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy#L399)), unlike Flink, which supports the latest three Flink versions.

Comment (Hannah-Jiang) [reply release/src/main/scripts/build_release_candidate.sh]: How about in future? In addition, I think users should know which spark version is supported even if only one version is supported.

**Comment by ibzib** [comment]
@Hannah-Jiang are we expecting to migrate Apache's repo to starting with 2.20?

**Comment by Hannah-Jiang** [comment]
> @Hannah-Jiang are we expecting to migrate Apache's repo to starting with 2.20?

Yes!

File: release/src/main/scripts/build_release_candidate.sh (hunk 1 above)
Comment (ibzib) [reply release/src/main/scripts/build_release_candidate.sh]: Minor clarification to the above: Beam Spark runner supports Spark's latest _feature_ release, i.e. `2.4.*`.

I think the plan is to create a separate module when moving to Spark 3 ([BEAM-9282](https://issues.apache.org/jira/browse/BEAM-9282)). @iemejia knows more

**Comment by ibzib** [comment]
> Yes!

Great. In that case, shall we merge this first, and then I'll let you patch it into #10612?

**Comment by Hannah-Jiang** [comment]
> > Yes!
> 
> Great. In that case, shall we merge this first, and then I'll let you patch it into #10612?

Yes, no problem.

File: release/src/main/scripts/build_release_candidate.sh (hunk 1 above)
Comment (Hannah-Jiang) [reply release/src/main/scripts/build_release_candidate.sh]: Then let’s add it to documentation some where. If I am a spark user, I would really wonder which version is being supported.

Comment (ibzib) [reply release/src/main/scripts/build_release_candidate.sh]: It is documented on the Spark runner page: https://beam.apache.org/documentation/runners/spark/

**Comment by ibzib** [comment]
Flake is unrelated ([BEAM-8912](https://issues.apache.org/jira/browse/BEAM-8912))

File: release/src/main/scripts/build_release_candidate.sh (hunk 1 above)
Comment (Hannah-Jiang) [reply release/src/main/scripts/build_release_candidate.sh]: Just noticed that `./gradlew` is repeated here as well.

Comment (ibzib) [reply release/src/main/scripts/build_release_candidate.sh]: 😰 fixed

**Comment by Hannah-Jiang** [comment]
LGTM

**Pull request merged**